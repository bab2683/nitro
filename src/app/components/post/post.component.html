<div class="post">
  <div class="post__content">
    <h2>{{ post.author }}</h2>
    <p class="post__date">
      {{ post.location }} - {{ post.date | date: "dd/MM/yyyy" }}
    </p>
    <p class="post__text">
      {{ post.text }}
    </p>
    <app-button
      class="post__edit-cta"
      [open]="editing"
      (clicked)="editing = !editing"
    >
      {{ editing ? "Close" : "Edit" }}</app-button
    >

    <div class="post__edit" *ngIf="editing">
      <form [formGroup]="form" (ngSubmit)="onSubmit()">
        <div class="post__edit__controls">
          <div *ngFor="let field of editableFields">
            <label class="post__edit__label" [for]="field">{{ field }}</label>
            <input
              class="post__edit__input"
              type="text"
              [formControlName]="field"
            />
            <p
              class="post__edit__error"
              *ngIf="
                !form.valid && form.dirty && submitted && !form.get(field).valid
              "
            >
              {{ field }} cannot be empty
            </p>
          </div>
        </div>
        <app-button class="post__edit__cta" (clicked)="onSubmit()"
          >Save changes</app-button
        >
      </form>
    </div>
  </div>
</div>
