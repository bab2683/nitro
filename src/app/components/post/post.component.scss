@import "common";

.post {
  &__cta {
    background-color: color("accent");
    border-radius: calc-spacing();
    color: color("white");
    display: block;
    padding: calc-spacing(2);
    transition: background-color 200ms ease-in-out, color 200ms ease-in-out;

    &:hover {
      background-color: color("accent-light-20");
    }
  }

  &__date {
    font-size: calc-fz(0.8);
  }

  &__text {
    margin-top: calc-spacing(2);
  }

  &__content {
    background-color: color("grey");
    border-radius: calc-spacing(1);
    color: color("white");
    margin: calc-spacing(2) 0 0 calc-spacing(2);
    padding: calc-spacing(2);
  }
}
