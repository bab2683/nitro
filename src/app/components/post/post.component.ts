import { Component, Input } from '@angular/core';

import { ParsedPost } from '@models';

@Component({
  selector: 'app-post',
  templateUrl: './post.component.html',
  styleUrls: ['./post.component.scss'],
})
export class PostComponent {
  @Input() public post: ParsedPost;
}
