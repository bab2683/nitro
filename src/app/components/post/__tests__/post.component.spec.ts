import { NO_ERRORS_SCHEMA } from '@angular/compiler/src/core';
import { DebugElement } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormBuilder, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';

import { PostComponent } from '../post.component';

describe('PostComponent', () => {
  let component: PostComponent;
  let fixture: ComponentFixture<PostComponent>;
  let formBuilder: FormBuilder;
  let editButton: DebugElement;
  let saveChangesButton: DebugElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [PostComponent],
      providers: [FormBuilder],
      imports: [FormsModule, ReactiveFormsModule],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(PostComponent);
    component = fixture.componentInstance;
    formBuilder = TestBed.get(FormBuilder);
    component.post = {
      author: 'Braulio',
      location: 'Dublin',
      id: 0,
      time: '1589642098034',
      text: '',
      summary: '',
      date: new Date(1589642098034),
      weekNumber: 21,
      open: false,
    };
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('ngOnInit', () => {
    it('should create the form', () => {
      jest.spyOn(formBuilder, 'group');

      component.ngOnInit();
      fixture.detectChanges();

      expect(formBuilder.group).toHaveBeenCalled();
    });
  });

  it('should match snapshot when edit panel is open', () => {
    editButton = fixture.debugElement.query(By.css('.post__edit-cta'));

    editButton.triggerEventHandler('clicked', null);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  describe('onSubmit', () => {
    beforeEach(() => {
      component.editing = true;

      fixture.detectChanges();
    });

    it('should save the values', () => {
      const authorField: DebugElement = fixture.debugElement.query(
        By.css('.post__edit__input')
      );

      saveChangesButton = fixture.debugElement.query(
        By.css('.post__edit__cta')
      );

      jest.spyOn(component, 'onSubmit');
      authorField.nativeElement.value = 'Braulio Barahona';

      saveChangesButton.triggerEventHandler('clicked', null);

      fixture.detectChanges();
      expect(component.onSubmit).toHaveBeenCalled();
    });
  });
});
