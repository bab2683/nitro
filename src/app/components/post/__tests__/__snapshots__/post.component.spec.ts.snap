// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostComponent should match snapshot 1`] = `
<app-post
  editableFields={[Function Array]}
  editing="false"
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  post={[Function Object]}
  submitted="false"
>
  <div
    class="post"
  >
    <button
      class="post__cta"
      type="button"
    >
        
    </button>
    
  </div>
</app-post>
`;

exports[`PostComponent should match snapshot when post is opened 1`] = `
<app-post
  editableFields={[Function Array]}
  editing="false"
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  post={[Function Object]}
  submitted="false"
>
  <div
    class="post"
  >
    <button
      class="post__cta"
      type="button"
    >
        
    </button>
    
    <div
      class="post__content"
    >
      <h2>
        Braulio
      </h2>
      <p
        class="post__date"
      >
         Dublin - 16/05/2020 
      </p>
      <p
        class="post__text"
      >
          
      </p>
      <button
        class="post__edit-cta"
        type="submit"
      >
         Edit 
      </button>
      
    </div>
  </div>
</app-post>
`;
